credentials/local/local_test.go

```diff
@@ -0,0 +1,209 @@
+	conn, _ := net.Dial(network, lisAddr)
+	defer conn.Close()
+	clientAuthInfo, err := hs(conn, lisAddr)
+	if err != nil {
+		t.Fatal("Error on client while handshake.")
+		return nil, err
+	}
+	return clientAuthInfo, nil
+}
+
+type testServerHandleResult struct {
+	authInfo credentials.AuthInfo
+	err      error
+}
+
+// Server accepts a client's connection with local credentials.
+func serverHandle(t *testing.T, hs serverHandshake, done chan testServerHandleResult, lis net.Listener) {
+	serverRawConn, err := lis.Accept()
+	serverAuthInfo, err := hs(serverRawConn)
+	if err != nil {
+		t.Errorf("Server failed while handshake. Error: %v", err)
+		serverRawConn.Close()
+		done <- testServerHandleResult{authInfo: nil, err: err}
+	}
+	done <- testServerHandleResult{authInfo: serverAuthInfo, err: nil}
+}
+
+func ServerAndClientHandshake(t *testing.T, network, listAddr string) credentials.SecurityLevel {
+	done := make(chan testServerHandleResult, 1)
+	timeout := time.After(100 * time.Second)
+	lis, err := net.Listen(network, listAddr)
+	if err != nil {
+		if strings.Contains(err.Error(), "bind: cannot assign requested address") ||
+			strings.Contains(err.Error(), "socket: address family not supported by protocol") {
+			t.Skipf("no support for address %v", listAddr)
+		}
+		t.Fatalf("Failed to listen: %v", err)
+		return credentials.Invalid
+	}
+	go serverHandle(t, serverLocalHandshake, done, lis)
+	defer lis.Close()
+	clientAuthInfo, err := clientHandle(t, clientLocalHandshake, network, lis.Addr().String())
+	if err != nil {
+		t.Fatalf("Error at client-side: Error: %v", err)
+		return credentials.Invalid
+	}
+	select {
+	case <-timeout:
+		t.Fatal("Test didn't finish in time")
+		return credentials.Invalid
+	case serverHandleResult := <-done:
+		if serverHandleResult.err != nil {
+			t.Fatalf("Error at server-side: Error: %v", serverHandleResult.err)
+			return credentials.Invalid
+		}
+		clientLocal, _ := clientAuthInfo.(Info)
+		serverLocal, _ := serverHandleResult.authInfo.(Info)
+		clientSecLevel := clientLocal.CommonAuthInfo.SecurityLevel
+		serverSecLevel := serverLocal.CommonAuthInfo.SecurityLevel
+		if clientSecLevel != serverSecLevel {
+			t.Fatalf("client's AuthInfo contains %s but server's AuthInfo contains %s.", clientSecLevel.String(), serverSecLevel.String())
+			return credentials.Invalid
+		}
+		return clientSecLevel
+	}
+	return credentials.Invalid
+}
+
+func (s) TestServerAndClientHandshake(t *testing.T) {
+	testCases := []struct {
+		testNetwork string
+		testAddr    string
+		want        credentials.SecurityLevel
+	}{
+		{
+			testNetwork: "tcp",
+			testAddr:    "127.0.0.1:10000",
+			want:        credentials.NoSecurity,
+		},
+		{
+			testNetwork: "tcp",
+			testAddr:    "[::1]:10000",
+			want:        credentials.NoSecurity,
+		},
+		{
+			testNetwork: "tcp",
+			testAddr:    "localhost:10000",
+			want:        credentials.NoSecurity,
+		},
+		{
+			testNetwork: "unix",
+			testAddr:    "/tmp/grpc_fullstack_test",
```

Comment:
Doing something like
`testAddr: fmt.Sprintf("/tmp/grpc_fullstack_test/%s", time.Now().UnixNano()`
might be better to make sure that the socket file doesn't already exist.